src/Syntax/TimeAgnostic.hs

```diff
@@ -0,0 +1,539 @@
+at different steps in the AST refinement process. This way we don't have to
+update mirrored types, instances and functions in lockstep.
+
+Some terms in here are always 'Timed'. This indicates that their timing must
+*always* be explicit. For the rest, all timings must be implicit in source files
+(i.e. 'Untimed'), but will be made explicit (i.e. 'Timed') during refinement.
+-}
+
+module Syntax.TimeAgnostic (module Syntax.TimeAgnostic) where
+
+import Control.Applicative (empty)
+
+import Data.Aeson
+import Data.Aeson.Types
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Data.List (genericTake,genericDrop)
+import Data.Map.Strict (Map)
+import Data.String (fromString)
+import Data.Text (pack)
+import Data.Typeable
+import Data.Vector (fromList)
+
+import EVM.Solidity (SlotType(..))
+
+-- Reexports
+import Syntax.Timing  as Syntax.TimeAgnostic
+import Syntax.Untyped as Syntax.TimeAgnostic (Id, Interface(..), EthEnv(..), Decl(..))
+
+-- AST post typechecking
+data Claim t
+  = C (Constructor t)
+  | B (Behaviour t)
+  | I (Invariant t)
+  | S Store
+deriving instance Show (InvariantPred t) => Show (Claim t)
+deriving instance Eq   (InvariantPred t) => Eq   (Claim t)
+
+data Transition t
+  = Ctor (Constructor t)
+  | Behv (Behaviour t)
+  deriving (Show, Eq)
+
+type Store = Map Id (Map Id SlotType)
+
+-- | Represents a contract level invariant along with some associated metadata.
+-- The invariant is defined in the context of the constructor, but must also be
+-- checked against each behaviour in the contract, and thus may reference variables
+-- that are not present in a given behaviour (constructor args, or storage
+-- variables that are not referenced in the behaviour), so we additionally
+-- attach some constraints over the variables referenced by the predicate in
+-- the `_ipreconditions` and `_istoragebounds` fields. These fields are
+-- seperated as the constraints derived from the types of the storage
+-- references must be treated differently in the constructor specific queries
+-- (as the storage variables have no prestate in the constructor...), wheras
+-- the constraints derived from the types of the environment variables and
+-- calldata args (stored in _ipreconditions) have a uniform semantics over both
+-- the constructor and behaviour claims.
+data Invariant t = Invariant
+  { _icontract :: Id
+  , _ipreconditions :: [Exp Bool t]
+  , _istoragebounds :: [Exp Bool t]
+  , _predicate :: InvariantPred t
+  }
+deriving instance Show (InvariantPred t) => Show (Invariant t)
+deriving instance Eq   (InvariantPred t) => Eq   (Invariant t)
+
+-- | Invariant predicates are either a single predicate without explicit timing or
+-- two predicates which explicitly reference the pre- and the post-state, respectively.
+-- Furthermore, if we know the predicate type we can always deduce the timing, not
+-- only vice versa.
+type family InvariantPred (t :: Timing) = (pred :: *) | pred -> t where
```

Comment:
Sure! A type family is a way to implement a type-level function. In this case we're saying that it's a function from the kind `Timing` (i.e. the input will be `Timed` or `Untimed`) to a predicate of kind `*` (i.e. a normal type). And with `| pred -> t` we're giving the type system a hint that not only will a timing give us a predicate type, we're also saying that every predicate type will uniquely determine a timing. It will try to verify this during type checking, and will then be able to use that information when checking other functions.

However I just tried to remove it and the entire project still type checks. I probably needed it for some experiment/attempt that later got scrapped. Maybe we should delete it and reintroduce it if we figure out that we need it in the future, to reduce noise? I don't mind keeping it around but I'm used to the syntax. Your call!

Fwiw, it's analogous to the more well-known functional dependencies in type classes. The canonical example there is multiplication:

```haskell
class Mult a b c where
  (*) :: a -> b -> c
```

If we specified a "can be multiplied" class like this (think numbers, but also vectors/matrices and vectors/scalars), we could give instances `Mult Vect Int Vect` *and* `Mult Vect Int Matrix`, which doesn't make sense at all. The type of the output is always completely dependent on the types of the inputs, so those two should exclude each other. Adding a functional dependency expresses this:

```haskell
class Mult a b c | a b -> c where
  (*) :: a -> b -> c
´´´

Reply:
We could also implement this using a GADT instead of a type family, like this:

```haskell
data InvPred (t :: Timing) where
  Single :: Exp Bool Untimed                 -> InvPred Untimed
  Double :: Exp Bool Timed -> Exp Bool Timed -> InvPred Timed
```

But then we would have to pattern match on `Single` and `Double` all the time.

Reply:
Makes sense! 

I'm fine with leaving things as they are now.